osxphotos/cli/sidecar.py

```diff
@@ -0,0 +1,144 @@
+            # in the past, so check if it exists and if so keep it
+            for filepath in export_results.skipped:
+                template_filename = _render_sidecar_filename(
+                    photo=photo,
+                    filepath=filepath,
+                    filename_template=filename_template,
+                    export_dir=export_dir,
+                    exiftool_path=exiftool_path,
+                )
+                if template_filename and pathlib.Path(template_filename).exists():
+                    verbose(
+                        f"Skipping existing sidecar file [filepath]{template_filename}[/]"
+                    )
+                    sidecar_results.sidecar_user_skipped.append(template_filename)
+
+        # write sidecar files for exported and missing files (and skipped if write_skipped)
+        files_to_process = export_results.exported + export_results.missing
+        if write_skipped:
+            files_to_process += export_results.skipped
+        for filepath in files_to_process:
+            template_filename = _render_sidecar_filename(
+                photo=photo,
+                filepath=filepath,
+                filename_template=filename_template,
+                export_dir=export_dir,
+                exiftool_path=exiftool_path,
+            )
+            if not template_filename:
+                raise click.BadOptionUsage(
+                    f"Invalid SIDECAR_FILENAME_TEMPLATE for --sidecar-template '{filename_template}'"
+                )
+
+            verbose(f"Writing sidecar file [filepath]{template_filename}[/]")
+            _render_sidecar_and_write_data(
+                template_file=template_file,
+                photo=photo,
+                template_filename=template_filename,
+                filepath=filepath,
+                strip_whitespace=strip_whitespace,
+                strip_lines=strip_lines,
+                dry_run=dry_run,
+            )
+            sidecar_results.sidecar_user_written.append(template_filename)
+
+    return sidecar_results
+
+
+def _render_sidecar_filename(
+    photo: PhotoInfo,
+    filepath: str,
+    filename_template: str,
+    export_dir: str,
+    exiftool_path: str,
+):
+    """Render sidecar filename template"""
+    render_options = RenderOptions(export_dir=export_dir, filepath=filepath)
+    photo_template = PhotoTemplate(photo, exiftool_path=exiftool_path)
+    template_filename, _ = photo_template.render(
+        filename_template, options=render_options
+    )
+    template_filename = template_filename[0] if template_filename else None
+
+    return template_filename
+
+
+def _render_sidecar_and_write_data(
+    template_file: str,
+    photo: PhotoInfo,
+    template_filename: str,
+    filepath: str,
+    strip_whitespace: bool,
+    strip_lines: bool,
+    dry_run: bool,
+):
+    """Render sidecar template and write data to file"""
+    sidecar = Template(filename=template_file)
```

Comment:
When I was looking through the templating code before, [I noticed you only compiled the built-in template once](https://github.com/RhetTbull/osxphotos/blob/02b6698c80903701cb56f6f8c50f1f3687be55dc/osxphotos/photoexporter.py#L2185-L2197). I'm assuming this takes a non-trivial amount of time so probably doesn't need to happen for every photo. Should there be a cache of `template_file` => compiled `Template`.

Reply:
Great idea. I'll add a cache.